src/error/Error.h

```diff
@@ -11,29 +11,53 @@
 
 #include "error/Detail.h"
 #include "error/forward.h"
+#include "mem/PoolingAllocator.h"
 
 #include <iosfwd>
+#include <vector>
+
+/// zero or more details of a single error
+using ErrorDetails = std::vector<ErrorDetailPointer, PoolingAllocator<ErrorDetailPointer> >;
+
+/// prints all given details separated by '+';
+/// if no details were given, prints nothing
+std::ostream &operator <<(std::ostream &, const ErrorDetails &);
 
 /// a transaction problem
 class Error {
 public:
     Error() = default;
     Error(const err_type c): category(c) {} ///< support implicit conversions
-    Error(const err_type c, const ErrorDetailPointer &d): category(c), detail(d) {}
+    Error(const err_type c, const ErrorDetailPointer &d): Error(c) { update(d); }
 
     explicit operator bool() const { return category != ERR_NONE; }
 
     /// if necessary, stores the given error information (if any)
     void update(const Error &);
 
-    /// convenience wrapper for update(Error)
-    void update(const err_type c, const ErrorDetailPointer &d) { update(Error(c, d)); }
+    /// if necessary, stores the given error information (if any);
+    /// more convenient and faster alternative to calling update(Error(c, d))
+    void update(err_type, const ErrorDetailPointer &);
+
+    /// if necessary, stores the given error category (if any), leaving details unchanged;
+    /// the first non-ERR_NONE category wins;
+    /// more convenient alternative to calling update(Error(c))
+    void update(err_type);
+
+    /// records an additional error detail (if any), leaving category unchanged
+    /// more convenient and faster alternative to calling update(Error(ERR_NONE, d))
+    /// \param detail either nil or a pointer to a new or an already known detail
+    void update(const ErrorDetailPointer &detail);
```

Comment:
We do not have to add this and the other new update() method. Earlier implementations did not have them. However, having these methods actually simplifies Error code (and improves caller code readability).